compiler/hash-parser/src/operator.rs

```diff
@@ -47,121 +56,217 @@ impl Operator {
+
+        match op {
+            // check if the operator is re-assignable and if so try to parse a equality operator.
+            // If we see this one then we can parse and then just convert the operator into the
+            // 'Eq' version.
+            Some(kind) => {
+                let assignable = match gen.peek_nth(consumed as usize) {
+                    Some(token) if kind.is_re_assignable() && token.has_atom(TokenAtom::Eq) => {
+                        consumed += 1;
+                        true
+                    }
+                    _ => false,
+                };
+
+                (Some(Operator { kind, assignable }), consumed)
+            }
+            None => (None, 0),
+        }
+    }
+}
+
+impl std::fmt::Display for OperatorKind {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            OperatorKind::EqEq => write!(f, "=="),
+            OperatorKind::NotEq => write!(f, "!="),
+            OperatorKind::BitOr => write!(f, "|"),
+            OperatorKind::Or => write!(f, "||"),
+            OperatorKind::BitAnd => write!(f, "&"),
+            OperatorKind::And => write!(f, "&&"),
+            OperatorKind::BitXor => write!(f, "^"),
+            OperatorKind::Exp => write!(f, "^^"),
+            OperatorKind::Gt => write!(f, ">"),
+            OperatorKind::GtEq => write!(f, ">="),
+            OperatorKind::Lt => write!(f, "<"),
+            OperatorKind::LtEq => write!(f, "<="),
+            OperatorKind::Shr => write!(f, "<<"),
+            OperatorKind::Shl => write!(f, ">>"),
+            OperatorKind::Add => write!(f, "+"),
+            OperatorKind::Sub => write!(f, "-"),
+            OperatorKind::Mul => write!(f, "*"),
+            OperatorKind::Div => write!(f, "/"),
+            OperatorKind::Mod => write!(f, "%"),
+            OperatorKind::As => write!(f, "as"),
         }
     }
+}
 
-    pub fn as_str(&self) -> &str {
+impl OperatorKind {
+    /// This returns if an operator is actually re-assignable. By re-assignable, this is in the sense
+    /// that you can add a '=' to mean that you are performing a re-assigning operation using the left
+    /// hand-side expression as a starting point and the rhs as the other argument to the operator.
+    /// For example, `a += b` is re-assigning because it means `a = a + b`.
+    pub(crate) fn is_re_assignable(&self) -> bool {
+        matches!(
+            self,
+            OperatorKind::BitOr
+                | OperatorKind::Or
+                | OperatorKind::BitAnd
+                | OperatorKind::And
+                | OperatorKind::BitXor
+                | OperatorKind::Exp
+                | OperatorKind::Shr
+                | OperatorKind::Shl
+                | OperatorKind::Add
+                | OperatorKind::Sub
+                | OperatorKind::Mul
+                | OperatorKind::Div
+                | OperatorKind::Mod
+        )
+    }
+
+    pub fn as_str(&self) -> &'static str {
```

Comment:
Maybe this isn't clear enough of a name. Maybe it should be called `descriptive_name` or something, because otherwise someone might think that .as_str() returns the actual operator (like the fmt::Display impl).

Reply:
Or we could go the route to actually implement the trait `fmt::Debug` and use that to convert into a descriptive name?